### ISSUE TITLE
File sorting is broken

### ISSUE THREAD

**Issue by KOPRajs** [xbmc/xbmc#22373]
### Describe the bug

While browsing my games collection in Kodi the files are not correctly sorted. See the screenshots.

### Screenshots 

![322738757_529984132427821_6935703135517252936_n](https://user-images.githubusercontent.com/13847813/210280740-f7647b19-3fab-4094-8b1b-e72f0f257aed.jpg)
![323100724_440905688125117_6559773641261650157_n](https://user-images.githubusercontent.com/13847813/210280755-1d207db1-ef66-4127-b020-101e2fd8e06d.jpg)

### Additional context or screenshots (if appropriate)

I've reproduced the bug on 2 different machines running LibreELEC and CoreELEC, but I can't reproduce it on Windows.

It affects all folders with multiple files using the same naming schema (ROM collection).

If I try to switch between ascending and descending order, I get different results everytime. Eventually I can get the correct order after several tries.

The same folders used to display correctly in Leia.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [X] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: LibreELEC 11
 - Kodi version: 20.0rc2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by KOPRajs** [comment]
I try to better explain the _random_ order that I'm seeing:

It is not completely random. There are always 2 groups of files and the sort order inside each group is correct. The distribution of the files between the 2 groups seems random.

It is like random files are skipped from the list and then they are appended to the end.

**Comment by vpeter4** [comment]
On which LibreELEC version you see this issue?

**Comment by KOPRajs** [comment]
Current LE nightly with Kodi 20.0rc2 and CoreELEC 20.0rc2.

**Comment by vpeter4** [comment]
So the common thing is arm? Because I think it doesn't happen on LibreELEC generic.

**Comment by KOPRajs** [comment]
Yes, both are ARMv8. I was not able to reproduce it on Windows either.

Also please look at this LE issue: https://github.com/LibreELEC/LibreELEC.tv/issues/7332

I believe it is not the same issue but there are similarities.

**Comment by KOPRajs** [comment]
I did some additional testing, here is what I've found:
1. It doesn't depend on the file size. The bug is reproducible even on empty files created by `touch`.
2. However it does seem to depend on the order in which the files were created. If the files were created in alphabetical order, then Kodi shows the folder correctly.

You can check the order of the inodes by running `find .` in the folder. If the result is already sorted, then Kodi shows the folder sorted as well.

Here is a code snippet that creates the files on which the bug can be reproduced on my system:
```
touch \
"./Tomb Raider (USA) (v1.6) (Track 53).bin" \
"./Tomb Raider (USA) (v1.6) (Track 15).bin" \
"./Tomb Raider (USA) (v1.6) (Track 45).bin" \
"./Tomb Raider (USA) (v1.6) (Track 10).bin" \
"./Tomb Raider (USA) (v1.6) (Track 57).bin" \
"./Tomb Raider (USA) (v1.6) (Track 06).bin" \
"./Tomb Raider (USA) (v1.6) (Track 13).bin" \
"./Tomb Raider (USA) (v1.6) (Track 47).bin" \
"./Tomb Raider (USA) (v1.6).cue" \
"./Tomb Raider (USA) (v1.6) (Track 39).bin" \
"./Tomb Raider (USA) (v1.6) (Track 56).bin" \
"./Tomb Raider (USA) (v1.6) (Track 24).bin" \
"./Tomb Raider (USA) (v1.6) (Track 35).bin" \
"./Tomb Raider (USA) (v1.6) (Track 50).bin" \
"./Tomb Raider (USA) (v1.6) (Track 48).bin" \
"./Tomb Raider (USA) (v1.6) (Track 41).bin" \
"./Tomb Raider (USA) (v1.6) (Track 05).bin" \
"./Tomb Raider (USA) (v1.6) (Track 31).bin" \
"./Tomb Raider (USA) (v1.6) (Track 40).bin" \
"./Tomb Raider (USA) (v1.6) (Track 09).bin" \
"./Tomb Raider (USA) (v1.6) (Track 36).bin" \
"./Tomb Raider (USA) (v1.6) (Track 52).bin" \
"./Tomb Raider (USA) (v1.6) (Track 38).bin" \
"./Tomb Raider (USA) (v1.6) (Track 51).bin" \
"./Tomb Raider (USA) (v1.6) (Track 27).bin" \
"./Tomb Raider (USA) (v1.6) (Track 21).bin" \
"./Tomb Raider (USA) (v1.6) (Track 07).bin" \
"./Tomb Raider (USA) (v1.6) (Track 29).bin" \
"./Tomb Raider (USA) (v1.6) (Track 49).bin" \
"./Tomb Raider (USA) (v1.6) (Track 55).bin" \
"./Tomb Raider (USA) (v1.6) (Track 18).bin" \
"./Tomb Raider (USA) (v1.6) (Track 44).bin" \
"./Tomb Raider (USA) (v1.6) (Track 17).bin" \
"./Tomb Raider (USA) (v1.6) (Track 34).bin" \
"./Tomb Raider (USA) (v1.6) (Track 32).bin" \
"./Tomb Raider (USA) (v1.6) (Track 08).bin" \
"./Tomb Raider (USA) (v1.6) (Track 54).bin" \
"./Tomb Raider (USA) (v1.6) (Track 42).bin" \
"./Tomb Raider (USA) (v1.6) (Track 12).bin" \
"./Tomb Raider (USA) (v1.6) (Track 30).bin" \
"./Tomb Raider (USA) (v1.6) (Track 02).bin" \
"./Tomb Raider (USA) (v1.6) (Track 33).bin" \
"./Tomb Raider (USA) (v1.6) (Track 43).bin" \
"./Tomb Raider (USA) (v1.6) (Track 23).bin" \
"./Tomb Raider (USA) (v1.6) (Track 20).bin" \
"./Tomb Raider (USA) (v1.6) (Track 22).bin" \
"./Tomb Raider (USA) (v1.6) (Track 11).bin" \
"./Tomb Raider (USA) (v1.6) (Track 46).bin" \
"./Tomb Raider (USA) (v1.6) (Track 04).bin" \
"./Tomb Raider (USA) (v1.6) (Track 28).bin" \
"./Tomb Raider (USA) (v1.6) (Track 26).bin" \
"./Tomb Raider (USA) (v1.6) (Track 25).bin" \
"./Tomb Raider (USA) (v1.6) (Track 37).bin" \
"./Tomb Raider (USA) (v1.6) (Track 16).bin" \
"./Tomb Raider (USA) (v1.6) (Track 03).bin" \
"./Tomb Raider (USA) (v1.6) (Track 14).bin" \
"./Tomb Raider (USA) (v1.6) (Track 01).bin" \
"./Tomb Raider (USA) (v1.6) (Track 19).bin"
```

Run this in an empty folder. Then check the output of `ls` vs. `find .`. While the `ls` should list the files sorted, the `find .` should list then in the same order that they were created. Then try to browse the folder in Kodi.

**Comment by vpeter4** [comment]
Can you try this CoreELEC image: https://mega.nz/file/3AxinDJJ#mWuPYpadUemp4LXRpfh9j6aeQSQB73BCVnuHMopvzj0

**Comment by KOPRajs** [comment]
Oh, sorry, I currently can't try the CoreELEC, I can only try it on LibreELEC (Rockchip RK3399).

**Comment by KOPRajs** [comment]
FYI: I've tested the same external disk on the Generic LE image and I can't reproduce it either, so it really seems to affect only ARM.

**Comment by chewitt** [comment]
@KOPRajs please test one of the RK3399 images from here: https://chewitt.libreelec.tv/testing/ .. the image(s) are current LE master with this PR/patch included (and odd versioning .. to make filenames static).

**Comment by KOPRajs** [comment]
So I've tried the [LibreELEC-RK3399.arm-10.88.0-rockpro64.img.gz](https://chewitt.libreelec.tv/testing/LibreELEC-RK3399.arm-10.88.0-rockpro64.img.gz) (2023-Jan-09 13:12) from the above link and I can't reproduce the Kodi sorting issue anymore, great work!

Please show us the magic PR as I'm really curios what was the problem.

Also I can confirm that the Busybox issue (Value too large for defined data type) while listing the directory in terminal is still there so it is definitely a separate issue.

**Comment by vpeter4** [comment]
Little magic in PR :)

**Comment by dl748** [comment]
I believe it'll still fail with any unicode character in a file. Dup of https://github.com/xbmc/xbmc/issues/20459

**Comment by vpeter4** [comment]
Can you make one folder with few empty files, zip it and upload somewhere?

**Comment by KOPRajs** [comment]
To add even more confusion: The pictures of the issue in the OP shows my CoreELEC box running 20.0rc2.
However, I can't reproduce the problem any more on that system since I've updated it to 20.0 final (the folders are the same). So whatever they've changed in CoreELEC between 20.0rc2 and 20.0 final, it seems to have fixed the issue even without the https://github.com/xbmc/xbmc/pull/22556.

**Comment by vpeter4** [comment]
CoreELEC included the patch temporary in CE20 release. Because why making release with known bugs where solutions are already ready :)

**Comment by KOPRajs** [comment]
Well, that explains it, thanks. I've actually thought that it might have been the case, so I've checked the https://github.com/CoreELEC/CoreELEC/tree/20.0-Nexus/packages/mediacenter/kodi/patches before, but I can't see the patch there.

**Comment by vpeter4** [comment]
You don't see it because patch file is in project folder.

**Comment by KOPRajs** [comment]
So I expect it should be listed here?
https://github.com/CoreELEC/xbmc/commits/5c67bb4e24920ebf3a5d45a084d5b78f8cc97f3c

I still can't seem to find it, but nevermind, I believe that it was included if you say that :-)

**Comment by KOPRajs** [comment]
Since this is fixed in 20.1, I'm going to close the issue. The https://github.com/xbmc/xbmc/issues/20459 seems to be a different issue.
